test/performance/README.md

```diff
@@ -0,0 +1,74 @@
+# Gimbal Performance Testing
+
+The Gimbal performance tests run an HTTP benchmarking tool (wrk2) to obtain
+performance metrics under different scenarios. The current version of the tests
+focus on the effects of the control plane on HTTP request/response performance.
+
+The performance tests assume that Gimbal has already been deployed, and that
+there is a backend cluster that can be used for deploying services and
+endpoints. In the ideal case, an additional cluster is available to run the HTTP
+benchmarking tool across multiple nodes. However, the backend cluster can be
+used for hosting backend workloads and benchmarking jobs.
+
+At the core, each test involves the following steps:
+
+- Create a namespace in all the clusters to run the test.
+- Make changes to the backend or gimbal cluster. For example, create services in
+  the backend cluster.
+- Wait until the changes have taken effect. For example, wait until all services
+  have been discovered.
+- Create a wrk2 job on the loadgen cluster.
+- Gather wrk2 logs once the job is complete.
+
+## Prerequisites
+
+- Gimbal cluster: Kubernetes cluster where all the Gimbal components are
+  running.
+- Backend cluster: Kubernetes cluster that has a corresponding discoverer
+  configured in the Gimbal cluster. Test services and endpoints will be created
+  in this cluster.
+- Load Generating cluster: Kubernetes cluster where wrk2 jobs will be deployed.
+- Gimbal endpoint: L3/L4 Load Balancer that is accessible from the Load
+  Generating cluster. This is the entrypoint to Gimbal.
+
+## Get Started
+
+### Label load generating nodes
+
+Nodes in the load generating cluster must be labeled with `workload=wrk2`. These
+are the nodes that will run the wrk2 job during the tests.
+
+### Cluster kubeconfigs
+
+Credentials to the clusters are provided using kubeconfig files. Currently,
+three kubeconfig files must be provided, one for each cluster. The path to the
+kubeconfig files is provided to the tests using the following flags:
+
+- `--gimbal-kubecfg-file`: Path to the kubeconfig file of the gimbal cluster
+- `--backend-kubecfg-file`: Path to the kubeconfig file of the backend cluster
+- `--loadgen-kubecfg-file`: Path to the kubeconfig file of the loadgen cluster.
+  Can be the same as `--backend-kubecfg-file`.
+
+### Gimbal URL
+
+The tests point wrk2 at a single URL, which usually is a L3/L4 load balancer
+that is in front of Gimbal. The URL is provided to the test using the
+`--gimbal-url` flag. For example, `http://gimbal.local`.
+
+### Run tests
+
+```sh
+make
+./gimbalbench \
+    --gimbal-kubecfg-file ${GIMBAL_KUBECONFIG} \
+    --backend-kubecfg-file ${BACKEND_KUBECONFIG} \
+    --loadgen-kubecfg-file ${LOADGEN_KUBECONFIG} \
+    --gimbal-url ${GIMBAL_URL} \
+    --concurrent-connections-test 250,500,1000,2500 \
+    --backend-services-test 10,25,50,100 \
+    --backend-endpoints-test 10,25,50,100
+```
+
+### Test results
+
+The wrk2 results are stored in the `logs` directory.
```

Comment:
Does each run get its own directory? i.e. can I run this over and over and have the logs for each run stored?